policy/policy-authority-data.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2015 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package policy
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+
+	blog "github.com/letsencrypt/boulder/log"
+
+	gorp "github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+)
+
+const whitelisted = "whitelist"
+const blacklisted = "blacklist"
+
+// DomainRule ...
+type DomainRule struct {
+	Host string `db:"host"`
+	Type string `db:"type"`
+}
+
+func reverseName(domain string) string {
+	labels := strings.Split(domain, ".")
+	for i, j := 0, len(labels)-1; i < j; i, j = i+1, j-1 {
+		labels[i], labels[j] = labels[j], labels[i]
+	}
+	return strings.Join(labels, ".")
+}
+
+// PolicyAuthorityDatabaseImpl enforces policy decisions based on various rule
+// lists
+type PolicyAuthorityDatabaseImpl struct {
+	log   *blog.AuditLogger
+	dbMap *gorp.DbMap
+}
+
+// NewPolicyAuthorityDatabaseImpl constructs a Policy Authority Database (and
+// creates tables if they are non-existent)
+func NewPolicyAuthorityDatabaseImpl(dbMap *gorp.DbMap) (padb *PolicyAuthorityDatabaseImpl, err error) {
+	logger := blog.GetAuditLogger()
+
+	dbMap.AddTableWithName(DomainRule{}, "ruleList").SetKeys(false, "Host")
+
+	padb = &PolicyAuthorityDatabaseImpl{
+		dbMap: dbMap,
+		log:   logger,
+	}
+
+	return padb, nil
+}
+
+// LoadRules loads the whitelist and blacklist into the database in a transaction
+// deleting any previous content
+func (padb *PolicyAuthorityDatabaseImpl) LoadRules(rules []DomainRule) error {
+	tx, err := padb.dbMap.Begin()
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	_, err = tx.Exec("DELETE FROM ruleList")
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	for _, r := range rules {
+		r.Host = reverseName(r.Host)
+		tx.Insert(&r)
+	}
+
+	err = tx.Commit()
+	return err
+}
+
+// DumpRules retrieves all DomainRules in the database so they can be written to
+// disk
+func (padb *PolicyAuthorityDatabaseImpl) DumpRules() ([]DomainRule, error) {
+	var dR []DomainRule
+	_, err := padb.dbMap.Select(&dR, "SELECT * FROM ruleList")
+	for _, r := range dR {
+		r.Host = reverseName(r.Host)
+	}
+
+	return dR, err
+}
+
+func (padb *PolicyAuthorityDatabaseImpl) checkBlacklist(host string) error {
+	var rule DomainRule
+	// Use lexical odering to quickly find blacklisted root domains
+	err := padb.dbMap.SelectOne(
+		&rule,
+		`SELECT * FROM ruleList WHERE :host >= host AND type = 'blacklist' ORDER BY host DESC LIMIT 1`,
```

Comment:
You can't `LIMIT 1` here, because you might get false positives. (e.g., `com.mail` for `com.mailthing`).  You need to apply the label-wise suffix match below to each candidate and return `nil` if any one matches.